expr/limit.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expr
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/nftables/binaryutil"
+	"github.com/mdlayher/netlink"
+	"golang.org/x/sys/unix"
+)
+
+// LimitType represents the type of the limit expression.
+type LimitType uint32
+
+// Imported from the nft_limit_type enum in netfilter/nf_tables.h.
+const (
+	LimitTypePkts     LimitType = unix.NFT_LIMIT_PKTS
+	LimitTypePktBytes LimitType = unix.NFT_LIMIT_PKT_BYTES
+)
+
+// Limit represents a rate limit expression.
+type Limit struct {
+	Type LimitType
+	Rate uint64
+	Over bool
+	// Possible values of unit:
+	// second, minute, hour, day, week
```

Comment:
This is a little vague. Are you trying to say that the granularity of `Unit` is whole seconds, and the maximum representable value is 1 week?

Reply:
nft will display the rules only for a second, a minute, an hour, a day and a week. This is an enum in the nft source code 
e.g.
```
udp dport 5006 limit rate 10000 bytes/minute
ip saddr 109.190.254.33 tcp dport 8080 limit rate 10/second
```

Now that the type is a time.Duration, you could set the duration to `2 * time.Second`, in that case, the rule will apply but nft will display `bytes/error`. I'm not sure if it works as expected though.

What you be better for the user, a better comment or an "enum" of type time.Duration representing the possible values ?

Reply:
Ah, now I understand. In that case, using time.Duration doesn’t make sense (users would need to convert/restrict them) after all, and I think we should just go for enum values with a custom type.

Reply:
I reintroduced the custom type with the possible values.